client/auth.go

```diff
@@ -0,0 +1,386 @@
+			AuthRoleListRequest: request,
+		},
+	}
+	resp, err := c.handleReq(&requestWithToken, true)
+	if err != nil {
+		return nil, err
+	}
+	res := resp.CommandResp.GetAuthRoleListResponse()
+	return res, err
+}
+
+// Deletes role.
+func (c *authClient) RoleDelete(request *xlineapi.AuthRoleDeleteRequest) (*xlineapi.AuthRoleDeleteResponse, error) {
+	requestWithToken := xlineapi.RequestWithToken{
+		Token: &c.token,
+		RequestWrapper: &xlineapi.RequestWithToken_AuthRoleDeleteRequest{
+			AuthRoleDeleteRequest: request,
+		},
+	}
+	resp, err := c.handleReq(&requestWithToken, false)
+	if err != nil {
+		return nil, err
+	}
+	res := resp.CommandResp.GetAuthRoleDeleteResponse()
+	res.Header.Revision = resp.SyncResp.Revision
+	return res, err
+}
+
+// Grants role permission
+func (c *authClient) RoleGrantPermission(request *xlineapi.AuthRoleGrantPermissionRequest) (*xlineapi.AuthRoleGrantPermissionResponse, error) {
+	if request.Perm == nil {
+		return nil, errors.New("Permission not given")
+	}
+
+	requestWithToken := xlineapi.RequestWithToken{
+		Token: &c.token,
+		RequestWrapper: &xlineapi.RequestWithToken_AuthRoleGrantPermissionRequest{
+			AuthRoleGrantPermissionRequest: request,
+		},
+	}
+	resp, err := c.handleReq(&requestWithToken, false)
+	if err != nil {
+		return nil, err
+	}
+	res := resp.CommandResp.GetAuthRoleGrantPermissionResponse()
+	res.Header.Revision = resp.SyncResp.Revision
+	return res, err
+}
+
+// Revokes role permission
+func (c *authClient) RoleRevokePermission(request *xlineapi.AuthRoleRevokePermissionRequest) (*xlineapi.AuthRoleRevokePermissionResponse, error) {
+	requestWithToken := xlineapi.RequestWithToken{
+		Token: &c.token,
+		RequestWrapper: &xlineapi.RequestWithToken_AuthRoleRevokePermissionRequest{
+			AuthRoleRevokePermissionRequest: request,
+		},
+	}
+	resp, err := c.handleReq(&requestWithToken, false)
+	if err != nil {
+		return nil, err
+	}
+	res := resp.CommandResp.GetAuthRoleRevokePermissionResponse()
+	res.Header.Revision = resp.SyncResp.Revision
+	return res, err
+}
+
+// Send request using fast path
+func (c authClient) handleReq(req *xlineapi.RequestWithToken, useFastPath bool) (*ProposeResponse, error) {
+	proposeId := c.generateProposeId()
+	cmd := xlineapi.Command{Request: req, ProposeId: proposeId}
+
+	if useFastPath {
+		res, err := c.curpClient.propose(&cmd, true)
+		return res, err
+	} else {
+		res, err := c.curpClient.propose(&cmd, false)
+		if err != nil {
+			return res, err
+		} else {
+			if res != nil && res.SyncResp == nil {
+				panic("syncResp is always Some when useFastPath is false")
+			}
+			return res, err
+		}
+	}
+}
+
+// Generate hash of the password
+func (c authClient) hashPassword(password []byte) string {
+	// Generate random salt
+	salt := make([]byte, 16)
+	_, err := rand.Read(salt)
+	// This doesn't seems to be fallible
+	if err != nil {
+		panic(err)
+	}
+	// Generate a hash of the password using PBKDF2
+	hashed := pbkdf2.Key(password, salt, 4096, 32, sha256.New)
```

Comment:
Using `const` to define 4096 and 32 would be better.

Reply:
Thank you for your feedback. I've updated the pull request based on your suggestions.